### DISCUSSION TITLE
Minify jss server side styles

### DISCUSSION THREAD

**Discussion by modbender** [vercel/next.js#17067] [Help] [answered]
Hey I'm using this example:
https://github.com/mui-org/material-ui/tree/master/examples/nextjs
But the styles in `<style id="jss-server-side"> CSS </style>` is not minimized, even in production build. Any way to minimize it?

.babelrc
```json
{
	"presets": ["next/babel"],
	"plugins": [
		["styled-components", {
			"ssr": true
		}]
	]
}
```

.next.config.js
```js
const withPlugins = require("next-compose-plugins");
const withCSS = require("@zeit/next-css");
const withSass = require("@zeit/next-sass");
const withFonts = require("next-fonts");
const withTM = require("next-transpile-modules")([
  "@material-ui/core",
  "@material-ui/icons",
  "@material-ui/styles",
  "@material-ui/lab",
]);
const withWorkers = require("@zeit/next-workers");

module.exports = withPlugins(
  [withTM, withCSS, withSass, withFonts, withWorkers],
  {
    poweredByHeader: false,
  }
);

```

Also I do have `NODE_ENV=production` in `.env.production`.

**Comment by shawwal** [comment]
same thing happen to me. Can't find any solution yet.

**Comment by chevaliermk** [comment]
for me same issue. everything inside `<style id="jss-server-side"></style>` isn't minified

**Comment by boostup** [comment]
In the NExtJS docs, here they make a reference to the `next.config.js` file to customizing plugins: https://nextjs.org/docs/advanced-features/customizing-postcss-config#customizing-plugins

So, your might wanna try this up: https://cssnano.co/docs/getting-started/#using-postcss-cli

The link, being PostCSS config...

---

Let me know if this works out for you guys, 'cause I'm about to start a Material UI + NextJS project myself, and while looking for potential issues with JSS and NextJS, I found the issue you're reporting.

What I'm wondering about though is:  is the **minimal JSS only** being served, just like the other default CSS modules supported out of the box by NextJS, as mentioned in their doc: 

_In production, all CSS Module files will be automatically concatenated into many minified and code-split .css files. These .css files represent hot execution paths in your application, **ensuring the minimal amount of CSS is loaded for your application to paint.**_ 

--source: https://nextjs.org/docs/basic-features/built-in-css-support

**Reply by sombii** [reply]
Hello @boostup , did you find any solution for this?? Can you please post a solution if any?

**Comment by sombii** [comment]
Hello, was anyone able to fix this issue ??

**Comment by Amir-Mirkamali** [comment, marked as answer]
You must minify it yourself. 
I tried this with clean-css 
first:
```
npm install clean-css
```

and then you can use this code: 

``` javascript

import React from 'react';
import Document, {
  Html, Head, Main, NextScript,
} from 'next/document';
import { ServerStyleSheets, } from '@material-ui/core/styles';

const CleanCSS = require('clean-css');
/* eslint-enable global-require */
const cleanCSS = new CleanCSS();

export default class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head />
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

MyDocument.getInitialProps = async (ctx) => {
  const sheets = new ServerStyleSheets();
  const originalRenderPage = ctx.renderPage;
  ctx.renderPage = () => originalRenderPage({ enhanceApp: (App) => (props) => sheets.collect(<App {...props} />), });
  const initialProps = await Document.getInitialProps(ctx);
  let css = sheets.toString();
  if (css) {
    css = cleanCSS.minify(css).styles;
  }

  return {
    ...initialProps,
    styles: [
      ...React.Children.toArray(initialProps.styles),
      <style id="jss-server-side" key="jss-server-side">{css}</style>,
    ],
  };
};

```

**Reply by SaharEbrahimii** [reply]
Great ! that was my question too and your solution woks for me .Thanks sir.

**Reply by Shiva-Br** [reply]
It works.

**Reply by cibulka** [reply]
It does work! However to my understanding it has some downsides:

1. According to [Bundleophobia](https://bundlephobia.com/package/clean-css@5.1.3), the Clean-CSS package has 166KB! If the Amir-Mirkamali's solution would be used, this package needs to be included to `dependencies` (not `devDependencies`), which increases the initial app load.
2. The CSS minification happens every time the app is rendered (with the `Document` component), which is not exactly efficient.

I think that the more performant way would be to create SSR styles during the build and then load the built file to the `Document`.

**Reply by 1xo2** [reply]
Amir you are a beast
